### ISSUE TITLE
Add DRM overlay planes API

### ISSUE THREAD

**Issue by ghost** [swaywm/wlroots#859]
Separate from #193, this issue will track status of using overlay planes.

---

wlroots has migrated to gitlab.freedesktop.org. This issue has been moved to:

https://gitlab.freedesktop.org/wlroots/wlroots/-/issues/859

**Comment by ascent12** [comment]
There is already some half-assed support for it in the DRM backend (because we use primary and cursor planes), but it really needs to be fleshed out a bit more and actually exposed in a useful way.
I'll get around to them in my https://github.com/ascent12/drm_doc series, but for now I'll give a quick overview:

An overlay/cursor plane is a framebuffer that can be laid on top of the CRTC (primary plane) and used to composite images together. Think exactly how we draw a window normally, but this all happens in hardware, so it's much more efficient, and we don't need to damage/redraw whatever is under it.
Overlay and cursor planes are pretty similar, but are used for different purposes.

## Why they're good

It allows us to perform a lot of operations without needing to wake the GPU up.
Less GPU means better battery life.

For example, we could perform an interactive drag for a floating window without needing to touch GL at all; we just move the plane.
We could implement things like layer-shell surfaces over the top of normal content without needing to damage content beneath it.

## Why they're so annoying to use

Due to the wide variation of hardware support, there is a lot of shit you have to deal with.
- There is no guarantee for the number of planes present
- They may only work with particular CRTCs (think wlr_output)
- They may or may not support hardware rotation
- They may or may not support Z-ordering

The "legacy" API for planes is complete crap and is unsynchronised with the pageflip, which would lead to horrible tearing. Weston has some weird shit they do to get around this, but I'd just limit this API to the atomic backend.

This is also going to make our damage tracking even more complicated.

## The API

I can't think of sane way to "wrap" this API for library users without removing a lot of the power and usefulness of planes. Perhaps we just expose them very close to how they are, and just let the user deal with the complexity.
It's a purely optional thing, so if they don't want to deal with it, they don't have to.

There is also going to be some interaction with #774.

**Comment by SirCmpwn** [comment]
>I can't think of sane way to "wrap" this API for library users without removing a lot of the power and usefulness of planes. Perhaps we just expose them very close to how they are, and just let the user deal with the complexity.

This idea is bourne of ignorance and probably stupid but I imagined having wlr_output learn the ability to manage planes (perhaps on wayland taking advantage of subsurfaces), via some functions to attempt to make new planes and some other functions to make them current and suchwhat.

**Comment by ascent12** [comment]
Hmm, I never thought about wayland subsurfaces when it comes to this. I could sort of see how that may work. Maybe if you took it far enough, you could pass off the real compositing to the parent compositor, and need to actually draw very little ourselves, but it would make input more complicated.

Basically, there would be a function `wlr_output_get_plane` which would get a plane __if it can__, and fail otherwise.
- X11 and Headless will always return NULL
- Non-atomic DRM will always return NULL (unless we choose to implement the same Weston hacks)
- Atomic DRM will return planes as long as it can (possibly none)
- Wayland may be able to return a valid plane indefinitely

Some extra restrictions on Atomic DRM
- If Z-ordering is not supported, we will only return a max of 1 plane per output, even if more are supported. The unspecified plane ordering is just not useful.
- By assigning a plane to a wlr_output, we may take away the ability for another wlr_output to have it. As such, getting the max number of planes for an output is not obvious, and could be extremely misleading. So we shouldn't implement a function for that; users just keep trying to take new planes until it fails.
- For the same reason above, planes shouldn't be considered a long-lived object that would last the entire lifetime of the wlr_output. They should be gotten and released frequently, as needed.
- When we hotplug an output, we may reshuffle the CRTCs used by each output, so we can get a modeset that works. We try as hard as we can not to, but sometimes it's unavoidable. This can change the set of supported planes for each wlr_output. So we may need to support the ability to revoke planes at any time.

**Comment by SirCmpwn** [comment]
Should we do soft planes like we do soft cursors if the hardware/backend/whatever isn't up to snuff?

**Comment by ascent12** [comment]
That kind of defeats the entire purpose of planes. It'll just punt all of the hard compositing problems into the renderer, instead of the user managing their windows themselves.

**Comment by SirCmpwn** [comment]
Alright, fair enough. Those problems will be hard, though.

**Comment by ascent12** [comment]
Also, just to bring things into context: Typical hardware (at least as far as I've seen) supports very few overlay planes.
Out of the 5 DRM devices I have access to, only 2 support overlay planes (both intel iGPU laptops), and they only support 1 overlay plane per CRTC.
We don't want to exclude any unknown hardware or hardware that may come, so we should be prepared for many more planes to be supported.

However, don't think of this like we can have every client sitting on its own plane.

**Comment by SirCmpwn** [comment]
I imagine it will only be practical to do fullscreen surfaces this way.

**Comment by ascent12** [comment]
Yeah, that would be the easiest one. There are a few extra little optimisations that would be possible even with a single overlay plane. We can just move clients between the main framebuffer (primary plane) to the overlay plane. So we could still get the efficient interactive drag, which is probably one of the most inefficient operations we can do when it comes to using the GPU.

**Comment by emersion** [comment]
Weston decided not to allow overlapping planes because the order of planes is ill-defined.

**Comment by emersion** [comment]
Related: support for underlay planes

>In general,the difference for underlay plane compared to overlay plane
is that the video data has to be  "underlaid" through a black
transparent quad rendered on the top plane(primary/RGB).
In other words, underlay plane has no individual existence w/o a primary
plane being rendered on top with black data or data size smaller than
that rendered on underlay plane.

**Comment by emersion** [comment]
Note that Weston makes no difference between a cursor plane and an overlay plane (apart from z-index). This means a window can be place on a cursor plane. After all, a cursor plane is just like an overlay plane with additional constraints. I think we could use the same approach to de-duplicate the code.

**Comment by emersion** [comment]
i915 patch to expose the zpos property: https://patchwork.freedesktop.org/series/58761/
